exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go

```diff
@@ -0,0 +1,220 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	runtimecatalog "sigs.k8s.io/cluster-api/internal/runtime/catalog"
+)
+
+// BeforeClusterCreateRequest is the request of the BeforeClusterCreate hook.
+// +kubebuilder:object:root=true
+type BeforeClusterCreateRequest struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// The cluster object the lifecycle hook corresponds to.
+	Cluster clusterv1.Cluster `json:"cluster"`
+}
+
+// BeforeClusterCreateResponse is the response of the BeforeClusterCreate hook.
+// +kubebuilder:object:root=true
+type BeforeClusterCreateResponse struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// CommonResponse contains Status and Message fields common to all response types.
+	CommonResponse `json:",inline"`
+
+	// RetryAfterSeconds when set to a non-zero value signifies that the hook
+	// will be called again at a future time.
+	RetryAfterSeconds int `json:"retryAfterSeconds"`
+}
+
+// BeforeClusterCreate is the runtime hook that will be called right before a Cluster is created.
+func BeforeClusterCreate(*BeforeClusterCreateRequest, *BeforeClusterCreateResponse) {}
+
+// AfterControlPlaneInitializedRequest is the request of the AfterControlPlaneInitialized hook.
+// +kubebuilder:object:root=true
+type AfterControlPlaneInitializedRequest struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// The cluster object the lifecycle hook corresponds to.
+	Cluster clusterv1.Cluster `json:"cluster"`
+}
+
+// AfterControlPlaneInitializedResponse is the response of the AfterControlPlaneInitialized hook.
+// +kubebuilder:object:root=true
+type AfterControlPlaneInitializedResponse struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// CommonResponse contains Status and Message fields common to all response types.
+	CommonResponse `json:",inline"`
+}
+
+// AfterControlPlaneInitialized is the runtime hook that will be called after the control plane is available for the first time.
+func AfterControlPlaneInitialized(*AfterControlPlaneInitializedRequest, *AfterControlPlaneInitializedResponse) {
```

Comment:
Should we also document here **why** the `AfterControlPlaneInitialized` hook is non-blocking? I did search for the reason in https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220330-topology-mutation-hook.md but did not find any.